File: Chain of responsibility/python/ChainOfResponsibility.py
import abc


class IWorker(metaclass=abc.ABCMeta):

    @abc.abstractmethod
    def set_next_worker(self, worker: 'IWorker') -> 'IWorker':
        pass


    @abc.abstractmethod
    def execute(self, command: str) -> str:
        pass


class AbsWorker(IWorker):

    def __init__(self):
        self.__next_worker : IWorker = None


    def set_next_worker(self, worker: 'IWorker') -> 'IWorker':
        self.__next_worker = worker
        return worker


    def execute(self, command: str) -> str:
        if self.__next_worker is not None:
            return self.__next_worker.execute(command)

        return ''


class Designer(AbsWorker):

    def execute(self, command: str) -> str:
        if command == 'design a house':
            return 'the designer executed the command:' + command
        return super().execute(command)



class Carpenters(AbsWorker):

    def execute(self, command: str) -> str:
        if command == 'laying a brick':
            return 'the carpenter executed the command:' + command
        return super().execute(command)


class FinishingWorker(AbsWorker):

    def execute(self, command: str) -> str:
        if command == 'glue wallpaper':
            return 'the interior decoration worker executed the command:' + command
        return super().execute(command)


def give_command(worker: IWorker, command: str):
    string : str = worker.execute(command)
    if string == '':
        print(command + 'no one knows how to do')
    else:
        print(string)


if __name__ == '__main__':

    designer = Designer()
    carpenters = Carpenters()
    finishing_worker = FinishingWorker()

    designer.set_next_worker(carpenters).set_next_worker(finishing_worker)

    give_command(designer, 'design a house')
    give_command(designer, 'laying a brick')
    give_command(designer, 'glue wallpaper')

    give_command(designer, 'conduct the transaction')

File: Chain of responsibility/java/ChainOfResponsibility.java
interface IWorker{
    
    IWorker setNextWorker(IWorker worker);
    String execute(String command);
}

abstract class AbsWorker implements IWorker{

    private IWorker nextWorker;

    public AbsWorker(){
        nextWorker = null;
    }

    @Override
    public IWorker setNextWorker(IWorker worker){
        nextWorker = worker;
        return worker;
    }

    @Override
    public String execute(String command){
        if (nextWorker != null)
            return nextWorker.execute(command);
        return "";
    }
}

class Designer extends AbsWorker{

    @Override
    public String execute(String command){
        if (command.equals("design a house")){
            return "the designer executed the command: " + command;
        }
        return super.execute(command);
    }
}

class Carpenters extends  AbsWorker{

    @Override
    public String execute(String command){
        if (command.equals("laying a brick")){
            return "the carpenter executed the command: " + command;
        }
        return super.execute(command);
    }
}

class FinishingWorker extends AbsWorker{

    @Override
    public String execute(String command){
        if (command.equals("glue wallpaper")){
            return "the interior decoration worker executed the command: " + command;
        }
        return super.execute(command);
    }
}

public class Main {

    public static void main(String[] args){
        Designer designer = new Designer();
        Carpenters carpenters = new Carpenters();
        FinishingWorker finishingWorker = new FinishingWorker();

        designer.setNextWorker(carpenters).setNextWorker(finishingWorker);

        GiveCommand(designer, "design a house");
        GiveCommand(designer, "laying a brick");
        GiveCommand(designer, "glue wallpaper");

        GiveCommand(designer, "conduct the transaction");
    }

    private static void GiveCommand(IWorker worker, String command){
        String str = worker.execute(command);
        if (str.equals(""))
            System.out.println("no one knows how to do");
        System.out.println(str);
    }
}